_drafts/commands.md

```diff
@@ -0,0 +1,82 @@
+---
+state: In Progress
+---
+
+# Commands
+
+"Commands" are an explicit interface for defining discrete pieces of functionality in a Hubot script. Initially, a command is a subset of the existing `robot.respond` interface. Over time, commands will start to gain more power and will be exposed through a command line interface or as native "slash commands" on services that support it, like Slack.
+
+## Motivation
+
+Most Hubot scripts have a command-ish interface, but the semantics of that interface are left to the whim of the script author. This can lead to confusing or conflicting scripts with unexpected behavior.
+
+Here are some of the problems with `robot.hear` and `robot.respond` for defining scripts:
+
+- listeners are registered with **regular expressions**, which can be cumbersome and error-prone, and encourage authors to design scripts with unpredictable interfaces.  
+- **Help documentation** must be manually defined because it cannot be generated from the definition of the listener, which leads to inconsistencies between implementation and documentation.
+- Many Hubot scripts would also be valuable if they were exposed through a command line interface or to other tools and services, but are currently **tightly coupled to a chat interface**.
+
+Here are some issues where similar features have been discussed:
+
+- https://github.com/github/hubot/issues/249
+- https://github.com/github/hubot/issues/873
+- https://github.com/slackapi/hubot-slack/issues/202
+- https://github.com/slackapi/hubot-slack/issues/223
+
+## Proposed solution
+
+A new API inspired by [Commander.js](https://github.com/tj/commander.js), a library with nice APIs for defining command-line interfaces, would be added on `robot.command`.
+
+```js
+robot
+  .command('cowsay <words ...>')
+  .description('make the cow say words')
+  .action(function(words) {
+    return require('cowsay').say(words.join(' '));
+  });
```

Comment:
I would recommend against a fluent API such as this. It is harder to test and for what you try to achieve it’s not well suited. I think a much clearer API would be this

```js
robot.command({
  name: 'cowsay',
  description: make the cow say words
  action: function (text) { // text being the text passed to the command
    return require('cowsay').say(text)
  }
})
```

It is just a JavaScript method without any magic, it will be simple to test and understand by users.